### PR TITLE
C++: Missing return-value check for scanf-like functions #1076

### PR DESCRIPTION
**Pull request by d10c**
- Update .gitignore for .vscode/*.log temporaries
- C++: Add test and placeholder query.
- C++: First working. We now prefer flagging the cases where the variable was initialized, as in real world cases we haven't seen it done safely.
- Add more (false-negative) MissingCheckScanf tests
- Add more MissingCheckScanf test cases

### CODE REVIEW

File: cpp/ql/test/query-tests/Critical/MissingCheckScanf/test.cpp [hunk 1]
```diff
@@ -94,22 +116,43 @@ int main()
...
+			use(i);
+		}
+	}
+
 	{
 		bool b;
 		int i;
 
-		b = scanf("%d", &i); // GOOD
+		b = scanf("%d", &i); // BAD [NOT DETECTED]: scanf can return EOF (boolifies true)
```
Comment (geoffw0) [inline cpp/ql/test/query-tests/Critical/MissingCheckScanf/test.cpp:135]: Well spotted!

File: cpp/ql/test/query-tests/Critical/MissingCheckScanf/test.cpp [hunk 2]
```diff
@@ -36,8 +36,30 @@ int main()
...
+	{
+		int i; // Reused variable
+
 		scanf("%d", &i); // BAD
 		use(i);
+
+		if (scanf("%d", &i) == 1) // GOOD
+		{
+			use(i);
+		}
```
Comment (geoffw0) [inline cpp/ql/test/query-tests/Critical/MissingCheckScanf/test.cpp:62]: Should we have a further copy of the first part (the `// BAD` bit) after the `// GOOD` part?  In a way `i` would have been "initialized" by the `// GOOD` part, but we would perhaps still consider the third part BAD for the same reason as line 39?

File: cpp/ql/test/query-tests/Critical/MissingCheckScanf/test.cpp [hunk 3]
```diff
@@ -94,22 +116,43 @@ int main()
...
+		b = scanf("%d", &i); // BAD [NOT DETECTED]
+
+		use(i);
+	}
+
 	{
 		int i, j;
 
-		if (scanf("%d %d", &i) >= 2) // GOOD
+		if (scanf("%d %d", &i) >= 2) // GOOD: `j` is not a scanf arg, so out of scope of MissingCheckScanf
```
Comment (geoffw0) [inline cpp/ql/test/query-tests/Critical/MissingCheckScanf/test.cpp:155]: I think I typo'd this, the intended test case was:
```
if (scanf("%d %d", &i, &j) >= 2) // GOOD
```

File: cpp/ql/src/Critical/MissingCheckScanf.ql [hunk 4]
```diff
@@ -11,9 +11,75 @@
...
+  private Instruction getAReuse() {
+    result = getNextInstruction() and
+    exists(Expr e | result.getAst() = e |
+      e instanceof ScanfOutput
+      or
+      e.getParent().(AddressOfExpr) instanceof ScanfOutput
+    )
+  }
+
+  private Instruction getNextInstruction() {
```
Comment (geoffw0) [inline cpp/ql/src/Critical/MissingCheckScanf.ql:73]: This needs a comment, it isn't obvious what a "next" instruction is.

File: cpp/ql/test/query-tests/Critical/MissingCheckScanf/test.cpp (hunk 3 above)
Comment (d10c) [reply cpp/ql/test/query-tests/Critical/MissingCheckScanf/test.cpp]: I added a separate test case for this :)

File: cpp/ql/test/query-tests/Critical/MissingCheckScanf/test.cpp (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/test/query-tests/Critical/MissingCheckScanf/test.cpp:64]: Seems reasonable.

File: cpp/ql/src/Critical/MissingCheckScanf.ql [hunk 5]
```diff
@@ -11,9 +11,75 @@
...
+ * - Combine into data flow tracking from out-arg to use,
+ *   with usage in _other_ scanf calls as barrier.
+ */
+
+class ScanfOutput extends Expr {
+  ScanfFunctionCall call;
+  int argNum;
+
+  ScanfOutput() {
+    this = call.getArgument(call.getFormatParameterIndex() + argNum) and
```
[the remark this answers is not in the document]
Comment (d10c) [reply cpp/ql/src/Critical/MissingCheckScanf.ql]: Resolved in a later commit.

File: cpp/ql/src/Critical/MissingCheckScanf.ql (hunk 4 above)
Comment (d10c) [reply cpp/ql/src/Critical/MissingCheckScanf.ql]: I renamed this to getASubsequentSameValuedInstruction in a later commit. Hopefully that's more self-explanatory :)

**Comment by d10c** [comment]
TODO: Remove a Cartesian Product! This query can run for hours and run out of space:

https://github.com/github/semmle-code/runs/7886119315?check_suite_focus=true#step:6:54

The offender is `getASubsequentSameValuedInstruction`.

**Comment by d10c** [comment]
MRVA top 10 run as of this commit: https://github.com/github/semmle-code/actions/runs/2914985782 
MRVA top 100 run as of this commit: https://github.com/github/semmle-code/actions/runs/2915212812

**Comment by d10c** [comment]
Superceded by squashed PR https://github.com/github/codeql/pull/10163

**Pull request closed without merging**